## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [36]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [37]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [38]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [39]:
# Use this space to do any additional research
# and familiarize yourself with the data.

#Product Analysis
average_price = df['unit_price'].mean()
average_cost = df['unit_cost'].mean()
average_weight = df['unit_weight'].mean()
total_units = df['qty'].sum()
cost_per_unit = average_cost / total_units
average_weight_per_unit = total_units / average_weight
print(f"Product Analysis (Average)\n"
      "---------------------------\n"
      f"Price of product:     ${round(average_price, 2)}\n"
      f"Cost of product:      ${round(average_cost, 2)}\n"
      f"Weight of product:    {round(average_weight, 2)} lbs\n"
      f"Cost per Pound:       ${round(average_cost / average_weight, 2)} per pound")


Product Analysis (Average)
---------------------------
Price of product:     $136.27
Cost of product:      $99.45
Weight of product:    5.0 lbs
Cost per Pound:       $19.87 per pound


In [57]:
top_sub_category = df.groupby('subcategory')[['qty', 'unit_cost', 'unit_price']].sum()
top_sub_category = top_sub_category.sort_values('qty', ascending=False)
top_sub_category.head()

qty  unit_cost  unit_price
subcategory                                      
stands             5521319  335039.54   448266.33
kitchen supplies   3293648  486536.33   673925.17
paper products     3107779  482533.01   687975.64
tables             3068302  526358.54   732488.91
bathroom supplies  2883256  687100.81   928394.99

In [58]:
# What three item categories had the most entries?
top_3_most_entries = df["category"].value_counts()
top_3_most_entries[0:3]

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [59]:
# For the category with the most entries, which subcategory had the most entries?
most_entries = 'consumables'
filtered_df = df[df['category'] == most_entries]
filtered_df['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [60]:
# Which five clients had the most entries in the data?
client_most_entries = df[['client_id']].value_counts().head(5)
client_most_entries

client_id
33615        220
66037        211
46820        209
38378        207
24741        207
Name: count, dtype: int64

In [61]:
# Store the client ids of those top 5 clients in a list.
top_5_list = df['client_id'].value_counts().head(5).index.tolist()
top_5_list

[33615, 66037, 46820, 38378, 24741]

In [62]:
# How many total units (the qty column) did the client with the most entries order order?
total_units = df.loc[df['client_id']== 33615, 'qty'].sum()
total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [63]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df[['unit_price', 'qty', 'subtotal']]


unit_price   qty   subtotal
0         1096.80   105  115164.00
1           24.95    21     523.95
2           13.52    39     527.28
3           36.42    29    1056.18
4          195.10    20    3902.00
...           ...   ...        ...
54634       83.13    33    2743.29
54635      206.59    47    9709.73
54636       65.66   475   31188.50
54637        1.48   112     165.76
54638        3.01  1031    3103.31

[54639 rows x 3 columns]

In [64]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['qty'] * df['unit_weight']
df['shipping_price'] = df['total_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
df[['qty', 'unit_weight', 'total_weight', 'shipping_price']]

qty  unit_weight  total_weight  shipping_price
0       105         7.50        787.50         5512.50
1        21         1.49         31.29          312.90
2        39         1.68         65.52          458.64
3        29         1.23         35.67          356.70
4        20        46.43        928.60         6500.20
...     ...          ...           ...             ...
54634    33         2.25         74.25          519.75
54635    47        11.70        549.90         3849.30
54636   475         4.16       1976.00        13832.00
54637   112        18.04       2020.48        14143.36
54638  1031         2.07       2134.17        14939.19

[54639 rows x 4 columns]

In [65]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = round((df['shipping_price'] + df['subtotal']) * 1.0925, 2)
df[['subtotal', 'shipping_price', 'total_price']]




subtotal  shipping_price  total_price
0      115164.00         5512.50    131839.08
1         523.95          312.90       914.26
2         527.28          458.64      1077.12
3        1056.18          356.70      1543.57
4        3902.00         6500.20     11364.40
...          ...             ...          ...
54634    2743.29          519.75      3564.87
54635    9709.73         3849.30     14813.24
54636   31188.50        13832.00     49184.90
54637     165.76        14143.36     15632.71
54638    3103.31        14939.19     19711.43

[54639 rows x 3 columns]

In [66]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['cost_of_each_line'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df[['first', 'last', 'unit_cost', 'qty', 'shipping_price', 'cost_of_each_line']]

first       last  unit_cost   qty  shipping_price  \
0         Donald    Harding     762.71   105         5512.50   
1        Tiffany      Myers      15.09    21          312.90   
2        Shannon     Watson       7.86    39          458.64   
3         Nathan      Baker      24.85    29          356.70   
4      Christina   Schwartz     108.17    20         6500.20   
...          ...        ...        ...   ...             ...   
54634     Lauren      Reese      51.60    33          519.75   
54635    Derrick      Moore     175.46    47         3849.30   
54636     Monica  Gutierrez      57.31   475        13832.00   
54637      Wanda    Solomon       1.22   112        14143.36   
54638     Dwayne       Bell       1.61  1031        14939.19   

       cost_of_each_line  
0               85597.05  
1                 629.79  
2                 765.18  
3                1077.35  
4                8663.60  
...                  ...  
54634            2222.55  
54635           12095.92  
54636           41054.25  
54637           14280.00  
54638           16599.10  

[54639 rows x 6 columns]

In [67]:
# Create a column for the profit of each line using line cost and line price
df['profit_of_each_line'] = df['total_price']  - df['cost_of_each_line']
df[['first', 'last', 'unit_price', 'cost_of_each_line', 'profit_of_each_line']]


first       last  unit_price  cost_of_each_line  \
0         Donald    Harding     1096.80           85597.05   
1        Tiffany      Myers       24.95             629.79   
2        Shannon     Watson       13.52             765.18   
3         Nathan      Baker       36.42            1077.35   
4      Christina   Schwartz      195.10            8663.60   
...          ...        ...         ...                ...   
54634     Lauren      Reese       83.13            2222.55   
54635    Derrick      Moore      206.59           12095.92   
54636     Monica  Gutierrez       65.66           41054.25   
54637      Wanda    Solomon        1.48           14280.00   
54638     Dwayne       Bell        3.01           16599.10   

       profit_of_each_line  
0                 46242.03  
1                   284.47  
2                   311.94  
3                   466.22  
4                  2700.80  
...                    ...  
54634              1342.32  
54635              2717.32  
54636              8130.65  
54637              1352.71  
54638              3112.33  

[54639 rows x 5 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [68]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]
def checkwork(order_ids):
    order_lines = df.loc[df['order_id'] == order_ids, 'total_price']
    return round(order_lines.sum(),2)

for x in order_ids:
    print(f"Order_id {x} had a total price of ${checkwork(x)}")




Order_id 2742071 had a total price of $152811.89
Order_id 2173913 had a total price of $162388.71
Order_id 6128929 had a total price of $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [69]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
def quantity_spent(client_id, column):
    top_5_spending = df.loc[df['client_id'] == client_id, column] 
    return round(top_5_spending.sum(), 2)

for x in top_5_list:
    print(f"{x} the total is {quantity_spent(x, 'shipping_price')}")


33615 the total is 1828984.89
66037 the total is 1395151.85
46820 the total is 1601448.84
38378 the total is 3429455.4
24741 the total is 5126448.37


In [70]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
def calculate_totals_for_top_clients(df, top_clients):
    results = []
    for client_id in top_clients:
        client_data = df[df['client_id'] == client_id]
        total_units_purchased = client_data['qty'].sum()
        total_shipping_price = client_data['shipping_price'].sum()
        total_revenue = client_data['total_price'].sum()
        total_profit = client_data['profit_of_each_line'].sum()
        results.append([client_id, total_units_purchased, total_shipping_price, total_revenue, total_profit])
    
    summary_df = pd.DataFrame(results, columns=['client_id', 'total units purchased', 'total shipping price', 'total revenue', 'total profit'])
    return summary_df

new_dataFrame = calculate_totals_for_top_clients(df, top_5_list)
new_dataFrame

client_id  total units purchased  total shipping price  total revenue  \
0      33615                  64313            1828984.89     8377308.52   
1      66037                  43018            1395151.85    10259514.79   
2      46820                  75768            1601448.84     9743794.36   
3      38378                  73667            3429455.40    12906550.87   
4      24741                 239862            5126448.37    82268892.02   

   total profit  
0    2201994.61  
1    3255031.81  
2    2736602.72  
3    3271829.89  
4   36579992.31

In [71]:
# Format the data and rename the columns to names suitable for presentation.
new_dataFrame = new_dataFrame.rename(columns={
    'client_id' : 'Client ID',
    'total units purchased' : 'Units Purchased'
}).copy()
# Define the money columns.
money_columns = new_dataFrame[['total shipping price', 'total revenue', 'total profit']]
# # Define a function that converts a dollar amount to millions.
def dollar_to_millions(x):
    return f"${x/1000000:.2f} M"

# Apply the currency_format_millions function to only the money columns. 
money_to_millions = money_columns.map(dollar_to_millions)
money_to_millions
# Rename the columns to reflect the change in the money format. 
renamed_df = money_to_millions.rename(columns={
    'total shipping price' : 'Shipping Price (in millions)',
    'total revenue':'Revenue (in millions)',
    'total profit':'Profit (in millions)',
})

renamed_df


Shipping Price (in millions) Revenue (in millions) Profit (in millions)
0                      $1.83 M               $8.38 M              $2.20 M
1                      $1.40 M              $10.26 M              $3.26 M
2                      $1.60 M               $9.74 M              $2.74 M
3                      $3.43 M              $12.91 M              $3.27 M
4                      $5.13 M              $82.27 M             $36.58 M

In [72]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df = renamed_df.sort_values("Profit (in millions)", ascending=False)
updated_df = sorted_df.reset_index(drop=True).copy()
columns_to_add = new_dataFrame[['Client ID', 'Units Purchased']]
final_df = pd.concat([columns_to_add, updated_df], axis=1).copy()
final_df


Client ID  Units Purchased Shipping Price (in millions)  \
0      33615            64313                      $5.13 M   
1      66037            43018                      $3.43 M   
2      46820            75768                      $1.40 M   
3      38378            73667                      $1.60 M   
4      24741           239862                      $1.83 M   

  Revenue (in millions) Profit (in millions)  
0              $82.27 M             $36.58 M  
1              $12.91 M              $3.27 M  
2              $10.26 M              $3.26 M  
3               $9.74 M              $2.74 M  
4               $8.38 M              $2.20 M

# Findings
----------------------------------
Based on the analysis, consumers purchased most out of the *Consumables* category, with the bathroom supplies sub-category being the leading item purchased in the Consumables category. 
Client ID 33615 is the companies top client. Bringing in a net profit of $36.58 Million dollars with 64, 313 units purchased. 